nibabel/openers.py

```diff
@@ -17,11 +17,39 @@
 GZIP_MAX_READ_CHUNK = 100 * 1024 * 1024  # 100Mb
 
 
+class BufferedGzipFile(gzip.GzipFile):
+    """GzipFile able to readinto buffer >= 2**32 bytes."""
+    def __init__(self, fileish, mode='rb', compresslevel=9, buffer_size=2**32-1):
+        super(BufferedGzipFile, self).__init__(fileish, mode=mode, compresslevel=compresslevel)
+        self.buffer_size = buffer_size
+
+        # Speedup for #209; attribute not present in in Python 3.5
+        # open gzip files with faster reads on large files using larger chunks
+        # See https://github.com/nipy/nibabel/pull/210 for discussion
+        if hasattr(self, 'max_chunk_read'):
+            gzip_file.max_read_chunk = GZIP_MAX_READ_CHUNK
+
+    def readinto(self, buf):
+        """Uses self.buffer_size to do a buffered read.
+
+        This works around a known issue in Python 3.5.
+        See https://bugs.python.org/issue25626"""
+        n_bytes = len(buf)
+        n_read = 0
+        max_read = 2 ** 32 - 1  # Max for unsigned 32-bit integer
+        while (n_read < n_bytes):
+            n_wanted = min(n_bytes - n_read, max_read)
+            print(n_read, n_wanted, n_bytes)
+            n_got = super(BufferedGzipFile, self).readinto(
+                buf[n_read:n_read + n_wanted])
```

Comment:
@matthew-brett this is not working; this slice apparently creates a different memory location to write to. Any idea about how to manage this?